docs/architecture/adr-053-state-sync-prototype.md

```diff
@@ -0,0 +1,198 @@
+}
+
+// Offers a snapshot to the application
+message RequestOfferSnapshot {
+    Snapshot snapshot = 1;
+}
+
+message ResponseOfferSnapshot {
+    bool accepted = 1;
+}
+
+// Fetches a snapshot chunk
+message RequestGetSnapshotChunk {
+    uint64 height = 1;
+    uint64 format = 2;
+    uint64 chunk = 3;
+}
+
+message ResponseGetSnapshotChunk {
+    SnapshotChunk chunk = 1;
+}
+
+// Applies a snapshot chunk
+message RequestApplySnapshotChunk {
+    SnapshotChunk chunk = 1;
+    bytes chain_hash = 2;
+    bool final = 3;
+}
+
+message ResponseApplySnapshotChunk {}
+```
+
+### Taking Snapshots
+
+Tendermint is not aware of the snapshotting process at all, it is entirely an application concern. The following guarantees must be provided:
+
+* **Periodic:** snapshots must be taken periodically, not on-demand, for faster restores, lower load, and less DoS risk.
+
+* **Deterministic:** snapshots must be deterministic, and identical across all nodes - typically by taking a snapshot at given height intervals.
+
+* **Consistent:** snapshots must be consistent, i.e. not affected by concurrent writes - typically by using a data store that supports versioning and/or snapshot isolation.
+
+* **Asynchronous:** snapshots must be asynchronous, i.e. not halt block processing and state transitions.
+
+* **Chunked:** snapshots must be split into chunks of reasonable size (on the order of megabytes), and each chunk must be verifiable against the chain Merkle root.
+
+* **Garbage collected:** snapshots must be garbage collected periodically.
+
+### Restoring Snapshots
+
+Nodes should have options for enabling state sync and/or fast sync, and be provided a trusted header hash for the light client.
+
+When starting an empty node with state sync and fast sync enabled, snapshots are restored as follows:
+
+1. The node checks that it is empty, i.e. that it has no state nor blocks.
+
+2. The node contacts the given seeds to discover peers.
+
+3. The node contacts a set of full nodes, and verifies the trusted block header using the given hash via the light client.
+
+4. The node requests available snapshots via `RequestListSnapshots`.
+
+5. The node iterates over all snapshots in reverse order by height and format until it finds one that satisfies both of the following conditions:
+
+    * The snapshot height's block is considered trustworthy by the light client (i.e. snapshot height is greater than trusted header and within unbonding period of the latest trustworthy block).
+
+    * The application accepts the `RequestOfferSnapshot` call.
+
+6. The node downloads chunks in parallel from multiple peers via `RequestGetSnapshotChunk`.
+
+7. The node passes chunks sequentially to the app via `RequestApplySnapshotChunk`, along with the root chain hash at the snapshot height for verification. If the chunk cannot be verified or applied, the application returns `ResponseException` and Tendermint tries refetching the chunk. The final chunk has `final: true`.
```

Comment:
Could you explain why chunks have to be applied sequentially? Because iAVL tree is order dependent?

Reply:
Yes, and also because I believe there were some security/DoS implications with nodes giving wrong indexes for chunks (e.g. chunk 3 being sent as chunk 1) and this not being verifiable until the entire tree had been rebuilt unless we verify them in order. I can't remember the details (see https://github.com/tendermint/tendermint/pull/3799), but it seems like a reasonable starting point anyway.